app/controllers/file_manager_controller.rb

```diff
@@ -0,0 +1,241 @@
+# Code for file manager adapted from: adrientoub/file-explorer
+require 'archive'
+require 'pathname'
+require 'mimemagic'
+
+class FileManagerController < ApplicationController
+  BASE_DIRECTORY = Rails.root.join('courses')
+  skip_before_action :set_course
+  skip_before_action :authorize_user_for_course
+  skip_before_action :update_persistent_announcements
+
+  def index
+    path = params[:path].nil? ? "" : params[:path]
+    new_url = "#{path}/"
+    parts = new_url.split('/')
+    if parts.empty?
+      @title = "File Manager"
+    else
+      @breadcrumbs << view_context.link_to("File Manager", file_manager_index_path)
+    end
+    parts.each_with_index do |part, index|
+      path = new_url.split('/').slice(0, index + 1).join("/")
+      @breadcrumbs << view_context.link_to(part, "/file_manager/#{path}")
+    end
+    absolute_path = check_path_exist(path)
+    if (File.directory?(absolute_path) && check_instructor(absolute_path)) || path == ""
+      populate_directory(absolute_path, new_url)
+      render 'file_manager/index'
+    elsif File.file?(absolute_path) && check_instructor(absolute_path)
+      if File.size(absolute_path) > 1_000_000 || params[:download]
+        send_file absolute_path
+      else
+        mime_type = MimeMagic.by_path(absolute_path).type
+        if mime_type.split('/').first == 'text'
+          @path = path
+          @file = absolute_path.read
+          render :file, formats: :html
+        else
+          send_file(absolute_path,
+                    filename: File.basename(absolute_path),
+                    disposition: 'attachment')
+        end
+      end
+    end
+  end
+
+  def upload
+    upload_file(params[:path].nil? ? "" : params[:path])
+  end
+
+  def delete
+    absolute_path = check_path_exist(params[:path])
+    return unless check_instructor(absolute_path)
+
+    parent = absolute_path.parent
+    raise "Unable to delete courses in the root directory." if parent == BASE_DIRECTORY
+
+    FileUtils.rm_rf(absolute_path)
+  end
+
+  def rename
+    absolute_path = check_path_exist(params[:relative_path])
+    if check_instructor(absolute_path)
+      parent = absolute_path.parent
+      if parent == BASE_DIRECTORY
+        flash[:error] = "Unable to rename courses in the root directory."
+      else
+        dir_name = File.dirname(params[:relative_path])
```

Comment:
`params[:relative_path]` has type `ActionController::Parameters`, best to cast it explicitly to string with `.to_s` which is what `File.dirname` expects

Reply:
Changed.